### DISCUSSION TITLE
Combination of two filters

### DISCUSSION THREAD

**Discussion by AlexandreGerault** [spatie/laravel-query-builder#967] [Ideas]
Hi!

I've been encoutering a use case on a project that seem not to be covered by the package. Indeed, we're using Laravel Query Builder since it is a great package and it increases the readibility of our query endpoints by far.

However, I'm writing a query where a filter actually depends on 2 filters. I want to filter the results by location: city and regions lets say. If I check city A, B and regions A, I would like to have results in cities A and B and also results in region A.

It's like a 
```php
$query->whereIn("city", ["A", "B"])->orWhereHas("region", fn ($q) => $q->where('name', 'B');
```

Thus the query string is looking like
```
/results?filter[cities]=A,B&filter[regions]=A
```

I just have no idea whether it could be handled by the package or not, and if it could, how would it looks like?

The workaround I have right now is to add 2 callback filters (so it allows the filters) but actually move the logic in a `when` method... As said, it looks like a workaround, and as it does not look like an issue, I thought it could be discussed.

```php
QueryBuilder::for(Result::query())
  ->allowedFilters([
    AllowedFilter::callback('cities'),
    AllowedFilter::callback('regions'),
  ])
  ->when(function () use ($request) {
    $hasCities = $request->has('filter') && array_key_exists('cities', $request->get('filter');
    $hasRegions= $request->has('filter') && array_key_exists('regions', $request->get('filter');
    
    return $hasCities || $hasRegions;
  }, function ($query) use ($request) {
    $query
      ->whereIn("city", request->get('filter')['cities'] ?? [])
      ->orWhereHas("region", fn ($q) => $q->where('name', $request->get('filter')['regions'] ?? []);
  })
  ->paginate();
```

Unless there's a way to handle it properly right now, I thought it could look like something like this:
```php
QueryBuilder::for(Result::query())
  ->allowedFilters([
    AllowedFilter::callback(['cities', 'regions'], function ($query, $cities, $regions) {
      $query
        ->whereIn("city", $cities)
        ->orWhereHas("region", fn ($q) => $q->where('name', $regions);
    }),
  ])
  ->paginate();
```

**Comment by pablosebastianr** [comment]
I was looking for something like this.

The situation in our application is slightly different. 

There are several defined filters the user can choose from, and some of them have a couple of other options for filtering on their own. These other options by themselves don't make sense so if the "main" filter is not set they do nothing

Your "empty" callbacks provided me the inspiration to solve it and this is what I came up with

```php 
        $rows = QueryBuilder::for(File::class)
            ->allowedFilters([
                AllowedFilter::callback('sent-from', fn(Builder $builder, $value) => $builder),
                AllowedFilter::callback('sent-to', fn(Builder $builder, $value) => $builder),
                AllowedFilter::callback('sent', function(Builder $builder, $value, $property) use ($request){

                    $filterKey = config('query-builder.parameters.filter', 'filter');

                    $from = $request->get($filterKey)['sent-from'] ?? null;
                    $to = $request->get($filterKey)['sent-to'] ?? null;

                    $builder->sentFilter($value, $from, $to);
                }),
            ])
            ->defaultSort('-updated_at')
            ->paginate($this->perPage());
```
I delegate to a scope inside the callback so the controller it's a bit cleaner.

It's a bit hacky but I don't think it looks that terrible. 
But honestly, it might be a big complication for little gain to have the ability to combine two filters since there are several types (exact, custom, scope, etc) and you still end up having to handle them in a callback.
